docs/graph.md

```diff
@@ -0,0 +1,714 @@
+    foo: int
+
+    async def run(
+        self,
+        ctx: GraphContext[MyState],
+    ) -> AnotherNode | End[int]:  # (2)!
+        if self.foo % 5 == 0:
+            return End(self.foo)
+        else:
+            return AnotherNode()
+```
+
+1. We parameterize the node with the return type (`int` in this case) as well as state.
+2. The return type of the `run` method is now a union of `AnotherNode` and `End[int]`, this allows the node to end the run if `foo` is divisible by 5.
+
+### Graph
+
+[`Graph`][pydantic_graph.graph.Graph] — The graph itself, made up of a set of nodes.
+
+`Graph` is generic in:
+
+* **state** the state type of the graph, [`StateT`][pydantic_graph.state.StateT]
+* **graph return type** the return type of the graph run, [`RunEndT`][pydantic_graph.nodes.RunEndT]
+
+Here's an example of a simple graph:
+
+```py {title="graph_example.py" py="3.10"}
+from __future__ import annotations
+
+from dataclasses import dataclass
+
+from pydantic_graph import BaseNode, End, Graph, GraphContext
+
+
+@dataclass
+class DivisibleBy5(BaseNode[None, int]):  # (1)!
+    foo: int
+
+    async def run(
+        self,
+        ctx: GraphContext,
+    ) -> Increment | End[int]:
+        if self.foo % 5 == 0:
+            return End(self.foo)
+        else:
+            return Increment(self.foo)
+
+
+@dataclass
+class Increment(BaseNode):  # (2)!
+    foo: int
+
+    async def run(self, ctx: GraphContext) -> DivisibleBy5:
+        return DivisibleBy5(self.foo + 1)
+
+
+fives_graph = Graph(nodes=[DivisibleBy5, Increment])  # (3)!
+result, history = fives_graph.run_sync(None, DivisibleBy5(4))  # (4)!
+print(result)
+#> 5
+# the full history is quite verbose (see below), so we'll just print the summary
+print([item.data_snapshot() for item in history])
+#> [DivisibleBy5(foo=4), Increment(foo=4), DivisibleBy5(foo=5), End(data=5)]
+```
+
+1. The `DivisibleBy5` node is parameterized with `None` as this graph doesn't use state, and `int` as it can end the run.
+2. The `Increment` node doesn't return `End`, so the `RunEndT` generic parameter is omitted, state can also be omitted as the graph doesn't use state.
+3. The graph is created with a sequence of nodes.
+4. The graph is run synchronously with [`run_sync`][pydantic_graph.graph.Graph.run_sync] the initial state `None` and the start node `DivisibleBy5(4)` are passed as arguments.
+
+_(This example is complete, it can be run "as is" with Python 3.10+)_
+
+A [mermaid diagram](#mermaid-diagrams) for this graph can be generated with the following code:
+
+```py {title="graph_example_diagram.py" py="3.10"}
+from graph_example import DivisibleBy5, fives_graph
+
+fives_graph.mermaid_code(start_node=DivisibleBy5)
+```
+
+```mermaid
+---
+title: fives_graph
+---
+stateDiagram-v2
+  [*] --> DivisibleBy5
+  DivisibleBy5 --> Increment
+  DivisibleBy5 --> [*]
+  Increment --> DivisibleBy5
+```
+
+## Stateful Graphs
+
+TODO introduce state
+
+TODO link to issue about persistent state.
```

Comment:
Not sure if you wanted to update this before merging, but here's a reminder in case you forgot. I'm okay either way.